tools/reviewer/web/src/app/dashboard/reviews-panel/diff/code-block/code-block.component.ts

```diff
@@ -0,0 +1,203 @@
+  }
+
+  // Highlitght code, split it by lines.
+  // Bootstrap method.
+  initLines(fileContent: string): void {
+    // TODO: detect language
+    const language = 'python';
+    const highlightedCode = this.highlightService.highlight(
+      fileContent,
+      language
+    );
+
+    const FileLines = fileContent.split('\n');
+    this.highlightedLines = highlightedCode.split('\n');
+    FileLines.forEach((code, i) => {
+      this.lines.push({
+        code: code,
+        hasPlaceholder: false,
+        height: 0,
+        highlightedCode: this.highlightedLines[i],
+        comments: [],
+        isChanged: false,
+        isCommentsVisible: false,
+      });
+    });
+  }
+
+  // Set which lines are changed
+  // Bootstrap method.
+  initChanges(changes: number[]): void {
+    if (!changes) {
+      return;
+    }
+    for (const lineNumber of changes) {
+      this.lines[lineNumber].isChanged = true;
+    }
+  }
+
+  // Add comments to lines
+  addComments(threads: Thread[]): void {
+    if (!threads) {
+      return;
+    }
+    for (const thread of threads) {
+      const i = thread.lineNumber;
+      this.lines[i].comments = thread.comments;
+      this.openCommentsBlock(i);
+    }
+  }
+
+  // Add a space for comments after the line inside highlighted code
+  addPlaceholder(i: number): void {
+    const div: HTMLDivElement = document.createElement('div');
+    div.style.display = 'inline-block';
+    div.style.height = this.lines[i].height + 'px';
+    div.innerText = ' ';
+    const placeholder = '\n' + div.outerHTML;
+    this.highlightedLines[i] = this.lines[i].highlightedCode + placeholder;
+
+    /*
+    'What is that?'
+
+    Short answer:
+      We need it because we have a dynamic height of the comment placeholders.
+
+    Long answer:
+      Comment component counts height of itself and sends it to parent.
+      The parent (Code block component) changes own view based on the height
+      after the view was checked, but before the component was initialized.
+      That causes the error:
+
+      'ERROR Error: ExpressionChangedAfterItHasBeenCheckedError:
+      Expression has changed after it was checked.'
+
+      https://stackoverflow.com/a/35243106
+      https://stackoverflow.com/a/46605947
+
+      To avoid the issue, we need to detect changes after changing the view.
+      Looks like Angular bug.
+    */
+    if (!this.changeDetectorRef['destroyed']) {
+      this.changeDetectorRef.detectChanges();
+    }
+  }
+
+  openCommentsBlock(i: number): void {
+    if (this.lines[i].hasPlaceholder) {
+      return;
+    }
+    this.lines[i].hasPlaceholder = true;
+    this.lines[i].isCommentsVisible = true;
+  }
+
+  closeCommentsBlock(i: number): void {
+    this.lines[i].hasPlaceholder = false;
+    this.lines[i].isCommentsVisible = false;
+    this.lines[i].height = 0;
+    this.highlightedLines[i] = this.lines[i].highlightedCode;
+  }
+
+  // Choose a color of changes highlighting
+  lineBackground(line: Line): string {
```

Comment:
You already have this method in `CodeBlockViewService.lineBackground()`, no?

Reply:
Right! Thanks. I guess I forgot to delete it from the component.